### ISSUE TITLE
Mypy cannot narrow the union type when boolean value is not directly evaluated in the assert statement

### ISSUE THREAD

**Issue by HiroIshida** [python/mypy#12676]
**Bug Report**
Mypy cannot narrow the union type when boolean value is not directly evaluated in the assert statement.

**To Reproduce**
(tested on the current master branch)
I saved the following code as `union_attr_error_without_context.py` and run
`mypy union_attr_error_without_context.py`
```python
from typing import Optional

class Model:
    var: int

class Example:
    var: Optional[Model] = None

    def is_var_set(self) -> bool:
        return self.var is not None

def func(ex: Example):
    #assert ex.var is not None
    assert ex.is_var_set()
    print(ex.var.var)
```

Note that the same error shows when I replace the assertion by below
```
boolean = ex.var is not None
assert boolean
```
Also note that, if I write 
```
assert ex.var is not None
```
then no error is shown.

**Expected Behavior**
No error is detected. Because by `assert ex.is_var_set()` ex.var is ensured that it is not None. So, `ex.var` is supposed to be not a union type.  

**Actual Behavior**
The following error shows:
```
union_attr_error_without_context.py:15: error: Item "None" of "Optional[Model]" has no attribute "var"  [union-attr]
```

**Your Environment**

- Mypy version used: cb6581a8f1f41bf784eebb6a1d2c8a32fb9b43f0 (head of the master branch)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files):
```
[mypy]
python_version = 3.8
show_error_codes = True
warn_unused_ignores = False
```
- Python version used: 3.8.10
- Operating system and version: ubuntu 20.04
